plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/incrementalProcessors.kt

```diff
@@ -0,0 +1,159 @@
+    ProcessingEnvironment by processingEnv {
+    override fun getFiler(): Filer = incFiler
+}
+
+internal class IncrementalFiler(private val filer: Filer) : Filer by filer {
+
+    internal var dependencyCollector: AnnotationProcessorDependencyCollector? = null
+
+    override fun createSourceFile(name: CharSequence?, vararg originatingElements: Element?): JavaFileObject {
+        val createdSourceFile = filer.createSourceFile(name, *originatingElements)
+        dependencyCollector!!.add(createdSourceFile.toUri(), originatingElements)
+        return createdSourceFile
+    }
+
+    override fun createClassFile(name: CharSequence?, vararg originatingElements: Element?): JavaFileObject {
+        val createdClassFile = filer.createClassFile(name, *originatingElements)
+        dependencyCollector!!.add(createdClassFile.toUri(), originatingElements)
+        return createdClassFile
+    }
+
+    override fun createResource(
+        location: JavaFileManager.Location?,
+        pkg: CharSequence?,
+        relativeName: CharSequence?,
+        vararg originatingElements: Element?
+    ): FileObject {
+        val createdResource = filer.createResource(location, pkg, relativeName, *originatingElements)
+        dependencyCollector!!.add(createdResource.toUri(), originatingElements)
+
+        return createdResource
+    }
+}
+
+internal class AnnotationProcessorDependencyCollector(private val runtimeProcType: RuntimeProcType) {
+    private val generatedToSource = mutableMapOf<File, File?>()
+    private var isFullRebuild = !runtimeProcType.isIncremental
+
+    internal fun add(createdFile: URI, originatingElements: Array<out Element?>) {
+        if (isFullRebuild) return
+
+        val generatedFile = File(createdFile)
+        if (runtimeProcType == RuntimeProcType.AGGREGATING) {
+            generatedToSource[generatedFile] = null
+        } else {
+            val srcFiles = getSrcFiles(originatingElements)
+            if (srcFiles.size != 1) {
+                isFullRebuild = true
+            } else {
+                generatedToSource[generatedFile] = srcFiles.single()
+            }
+        }
+    }
+
+    internal fun getGeneratedToSources(): Map<File, File?> = if (isFullRebuild) emptyMap() else generatedToSource
+    internal fun getRuntimeType(): RuntimeProcType {
+        return if (isFullRebuild) {
+            RuntimeProcType.NON_INCREMENTAL
+        } else {
+            runtimeProcType
+        }
+    }
+}
+
+private fun getSrcFiles(elements: Array<out Element?>): List<File> {
+    return elements.filterNotNull().mapNotNull { elem ->
+        var origin = elem
+        while (origin.enclosingElement != null && origin.enclosingElement !is PackageElement) {
+            origin = origin.enclosingElement
+        }
+        val uri = (origin as? Symbol.ClassSymbol)?.sourcefile?.toUri()?.takeIf { it.isAbsolute }
+        uri?.let { File(it) }
```

Comment:
@gavra0 I think we should deduplicate file paths here in case multiple elements originate from the same source file. This could happen, for example, if annotation processor generates multiple Kotlin top-level functions per element.

The place where Gradle handles this, for reference:
https://github.com/gradle/gradle/blob/e7dd3ba/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/IsolatingProcessingStrategy.java#L67-L71

Reply:
I agree. That is a case I did not consider, I'll upload a PR to address this.

Reply:
I've uploaded https://github.com/JetBrains/kotlin/pull/2361 to fix this.